### DISCUSSION TITLE
꼬인 커밋 내역 해결하기

### DISCUSSION THREAD

**Discussion by yeon-06** [woowacourse-study/2022-lv3-algorithm-study#76] [문제 같이 풀어요~]
이번에 커밋 내역이 꼬이신 분들이 많은 것 같은데요
제가 자주 사용하는 방법을 공유하고자 discussion 올립니다 :D

작업을 크게 나누자면 아래와 같습니다.
1. 커밋 내역을 옛날로 되돌리기
2. rebase 하기
3. 커밋 다시 하기
4. force push 하기

명령어로 좀 더 자세히 살펴볼게요~

- 현재 상황: origin의 `week2`에서 upstream의 `main`으로 PR을 올렸는데 충돌이 난다.
- 현재 작업 브랜치: origin의 `week2`

1. 이번주차 과제들 로컬에 백업해두기
2. git remote add upstream https://github.com/woowacourse-study/2022-lv3-algorithm-study.git
3. git reset --hard abb74cf4f2a6f24ad3aa33e2159d9fc4726695c9
  (-> 커밋 기록을 옛날로 돌리는 작업입니다. 해당 커밋 번호는 upstream 레포가 처음 만들어졌을때 생성한 커밋 번호입니다)
4. git rebase upstream/main
5. git pull upstream main
6. 1에서 백업한 과제들 다시 추가하고 커밋
7. git push -f origin week2

**Comment by kth990303** [comment]
좋은 글 감사합니다!!

커밋 내역을 옛날로 되돌리는 이유가 무엇인지 궁금해요. 현재 커밋 기록이 꼬인 분들이 수정하실 때, 커밋 기록을 지워야 하기 때문에 추가된 내용이라 생각되는데 맞을까요?? 4~7번까지만 해도 될 거 같아서 질문드려요!

참고로 저는 `fetch` -> `merge`/`rebase` 하고 작업한 소스코드를 넣어서 PR 날리고 있습니다. (`merge`를 하면 커밋 로그가 지난주차부터 존재하긴 하지만, files changed에는 현재주차 파일들만 들어가서 상관없어가지고..!) (대신, 로컬에 다른 분들의 소스코드까지 존재하긴 합니다 ㅎㅎ)

**Reply by yeon-06** [reply]
4~7만 해도 됩니다!
근데 이미 모든 작업 내역이 커밋된 후에 rebase를 하면 이전 커밋내역과 머지된 과정에 대해 추가로 커밋 하나가 생기더라고요
제가 rebase 명령어를 잘 몰라서 그런거일수도 있는데 파일이 충돌나서 수동으로 수정해야하는 경우도 생기고..
귀찮아서 그냥 옛날 커밋 기록으로 돌린 후 rebase하고 다시 커밋하는 편이에요 ㅎㅎ

**Reply by kth990303** [reply]
아, 제가 아마 작업한 소스코드를 커밋하기 전에 fetch -> upstream merge/rebase를 해서 그런 것 같아요. 저는 작업 전에 리베이스를 해버리고, 그 이후에는 upstream 수정을 최대한 안하고, 제 로컬 폴더에 제 소스코드를 복사해서 pr을 날리고 있어서 딱히 문제가 없는 것 같기도 한데... 확실친 않네요ㅋㅋㅋ

이번주차에는 merge를 했는데, 다음에는 rebase도 해보고, 이것저것 해보면서 git 공부해봐야겠어요 ㅎㅎ 감사합니다

**Reply by yeon-06** [reply]
맞아요 rebase를 이미 하신 상태라면 문제 없을거예요^ㅡ^b
제가 써둔 방법은 이미 모든게 끝난 뒤 뒤늦게 눈치챈.. 최악의 상황에 사용하는 방법입니다 ㅎㅎ
공부하면 블로그에 포스팅 올라오시나요? 기대하겠습니다😄

**Comment by hyeonic** [comment]
저는 애초에 새롭게 미션을 시작할 때 기존 브랜치는 삭제 한 뒤 main으로 부터 최신 커밋을 rebase 혹은 pull 해서 싱크를 맞춘 뒤 새롭게 브랜치를 생성합니다! 이렇게 되면 굳이 기존에 남긴 브랜치와 현재 브랜치 사이의 충돌을 최소화할 수 있어요!
